src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -765,15 +777,37 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 		bool queryOK = false;
 		bool failOnError = false;
 		int64 currentAffectedTupleCount = 0;
-		PGconn *connection = GetConnectionForPlacement(taskPlacement,
-													   isModificationQuery);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = GetNodeConnection(connectionFlags,
+															 taskPlacement->nodeName,
+															 taskPlacement->nodePort);
+		PGconn *connection = multiConnection->pgConn;
+
+		if (taskRequiresTwoPhaseCommit)
+		{
+			RemoteTransactionBeginIfNecessary(multiConnection);
+			MarkRemoteTransactionCritical(multiConnection);
+		}
```

Comment:
> I think we also should always use 2PC when transactions are in use, not just when reference tables are modified.

This makes a lot of sense. But, not sure whether to include in this PR. And, maybe let @ozgune to chime-in here? (Also, @marcocitus opened a similar issue on this #1032)